pkg/logql/log/filter.go

```diff
@@ -90,6 +89,83 @@ func (a andFilter) ToStage() Stage {
 	}
 }
 
+type andFilters struct {
+	filters []Filterer
+}
+
+// NewAndFilters creates a new filter which matches only if all filters match
+func NewAndFilters(filters []Filterer) Filterer {
+	var containsFilterAcc *containsAllFilter
+	regexpFilters := make([]Filterer, 0)
+	n := 0
+	for _, filter := range filters {
+		// Make sure we take care of panics in case a nil or noop filter is passed.
+		if !(filter == nil || filter == TrueFilter) {
+			switch c := filter.(type) {
+			case containsFilter:
+				// Start accumulating contains filters.
+				if containsFilterAcc == nil {
+					containsFilterAcc = &containsAllFilter{}
+				}
+
+				// Join all contain filters.
+				containsFilterAcc.Add(c)
+			case regexpFilter:
+				regexpFilters = append(regexpFilters, c)
+
+			default:
+				// Finish accumulating contains filters.
+				if containsFilterAcc != nil {
+					filters[n] = containsFilterAcc
```

Comment:
That's better. I think in the long term we may want to accumulate `not` too :)

Reply:
Yes. However, I'd like to land this PR first :slightly_smiling_face: